New license page.

TonicArtos opened the pull request:
## Description

This is a PR addressing #57226 - Proposal: New UI for Licenses Page.

This PR replaces the previous single panel license page with one that uses a master/detail flow (MDFlow) to display packages and their respective licenses.

##### Existing behaviour
The existing license page is a single list of all licenses collected from the packages linked in the application. Also, at the top of the list is an entry which displays some information about the application.

##### Issues with existing behaviour
1. It is hard to navigate.
Because all versions of a license must be displayed, some packages have many tens of similar entries in this list, often with little more than a date change. Should a user wish to see licenses for a particular package, they must hunt through the list to find the package and its licenses.
2. It is not polished.
The existing license page does not adjust for the size of the screen and can present extremely long strings of legalese running across the screen.

##### How this PR addresses the above (Changes)
The License Page API remains unchanged. The logic for processing the license data is kept largely the same. This PR changes how the licenses are displayed, by introducing a responsive UI using the master/detail UI pattern. For now I am calling it Master Detail Flow, or MDFlow.

MDFlow manifests as two layouts depending on the screen size. On small and medium displays, as determined by the breakpoints given by the Material Design Spec, MDFlow utilises a nested layout. On large displays, MDFlow uses a two panel (lateral) layout. MDFlow is implemented in this PR using a Navigator for the nested layout, and a Stack for the lateral layout. The master and detail views are built using builders. For the interactive component, detail pages are requested from the master view using a proxy obtained by a widget lookup on the build context; `MasterDetailFlow.of(context)`.

[Demo of new license page](https://tonicartos.github.io/flutter-new-license-page/#/)

**New License Page showing packages (small and medium displays)**
![Screenshot from 2020-05-20 01-03-13](https://user-images.githubusercontent.com/296142/82356357-028e1980-9a36-11ea-9d4a-579deeebec08.png)

**New License Page showing licenses for a package (small and medium displays)**
![Screenshot from 2020-05-20 01-03-20](https://user-images.githubusercontent.com/296142/82356385-0fab0880-9a36-11ea-9b2d-7164029f3f0e.png)

**New License Page (large displays)**
![Screenshot from 2020-05-20 01-02-57](https://user-images.githubusercontent.com/296142/82356395-13d72600-9a36-11ea-8b6d-e7476b123019.png)

**Existing License Page (all displays)**
![Screenshot from 2020-05-20 01-07-34](https://user-images.githubusercontent.com/296142/82356661-78928080-9a36-11ea-96bb-7b90524e5ed6.png)

## Related Issues

Resolves #57226
Resolves #18281

## Tests

Widget tests for the new UI remain to be updated.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [x] Design doc waved. See @Hixie https://github.com/flutter/flutter/pull/57588#issuecomment-631724000
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [x] I wrote a migration guide: **Migration should not be required**

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

TonicArtos committed:
Draft of new license UI.

TonicArtos committed:
Remove library directive.

TonicArtos committed:
No leading icon in lateral detail view.

TonicArtos committed:
Fix license header.

TonicArtos committed:
Fix missing defaults for about tile.

TonicArtos committed:
Show about program while package list is building.

TonicArtos committed:
Don't use builder for list of packages.

TonicArtos committed:
Fix fade anim not happening on load of packages.

TonicArtos committed:
Update AboutListTile control test.

Due to the new UI, a request for the pirate license must be made. This
change adds a tap request and short wait for the pirate license to
display.

TonicArtos committed:
Fix requiring app icon.

TonicArtos committed:
Update LicensePage control test.

Prior to this change, this test looks to see if two new licenses and
packages registered during this test are displayed by the license page.
It does this by checking for the non-existence and then existence of
widgets containing the correct content of these packages and licenses.

The change adds navigation of the new UI between checks for the license
content.

TonicArtos committed:
Update LicensePage control test with all props.

This change is the same as that for LicensePage control test.

TonicArtos commented:
@Hixie The new UI require changes to the tests in about_test.dart. The changes involve adding navigation of packages to show the license content, something the tests current assume is unneeded.

Anyway, is it okay to update these tests without writing a design doc?

TonicArtos committed:
Change unmounted tests to check packages called.

Changing to track packages called rather than paragraphs is because the
new UI only accesses the packages property of the license at first. Only
accessing the paragraphs property when the package is selected,
displaying the package licenses page.

TonicArtos committed:
Minor fixes to mdflow widget.

- Missing back button on narrow display.
- autoImplyLeading -> automaticallyImplyLeading

TonicArtos committed:
Only show back button if flow route can be popped.

TonicArtos committed:
Localisation for package detail text.

TonicArtos committed:
material_en for licensePackageDetailText.

TonicArtos committed:
Gen licensesPackageDetailText stubs from en.

TonicArtos committed:
Add fn declarations for localised plural.

- The guide in material/material_localizations.dart doesn't mention the
  extra work to support plurals.

Hixie commented:
> @Hixie The new UI require changes to the tests in about_test.dart. The changes involve adding navigation of packages to show the license content, something the tests current assume is unneeded.
> 
> Anyway, is it okay to update these tests without writing a design doc?

Yeah that should be fine. Thanks for checking.

TonicArtos committed:
Move mdflow into about.dart and make private.

TonicArtos committed:
Push canPop down to nested nav. Fix accessibility.

TonicArtos commented:
Okay, so I am looking for feedback now. I think these are the right people to tag. @HansMuller, @timsneath, @Hixie.

Hixie commented on packages/flutter/lib/src/material/about.dart:
```diff
@@ -556,49 +844,123 @@ class _LicensePageState extends State<LicensePage> {
+        MaterialLocalizations.of(context);
+    final double gutterSize = _getGutterSize(context);
+
+    if (widget.scrollController == null) {
```
there's a lot of duplicated content here, might be worth seeing if there's a way to factor it out so the branches are only the pieces that differ

Hixie commented on packages/flutter/lib/src/material/about.dart:
```diff
@@ -623,3 +985,662 @@ Widget _defaultApplicationIcon(BuildContext context) {
+typedef _ActionBuilder = List<Widget> Function(
+    BuildContext context,
+    _ActionLevel actionLevel,
+    );
```
these can all just be on one line

Hixie commented on packages/flutter/lib/src/material/about.dart:
```diff
@@ -472,6 +482,294 @@ class LicensePage extends StatefulWidget {
+    );
+  }
+
+  Widget _buildPackagesView(final BuildContext _, final bool isLateral) {
```
here and in other places where you have build methods, might be better to factor those out into their own widgets

Hixie commented:
Overall this looks great. I'll defer to @HansMuller or @goderbauer for more detailed review.

TonicArtos committed:
Fix code style and doc on builders.

- re: https://github.com/flutter/flutter/pull/57588#discussion_r428930769

TonicArtos committed:
Fix code style, <CR> after single line asignment.

TonicArtos committed:
Deduplicate code in package license page widget.

- re: https://github.com/flutter/flutter/pull/57588#discussion_r428930676

TonicArtos commented:
> Overall this looks great. I'll defer to @HansMuller or @goderbauer for more detailed review.

Okay. I'll follow your guidance once I their feedback. I don't want to be making changes over the top of what they may be reviewing.

TonicArtos committed:
Clean up nested navigator.

gspencergoog commented on packages/flutter_localizations/lib/src/l10n/material_en.arb:
```diff
@@ -72,6 +72,14 @@
     "description": "The title for the Flutter licenses page."
   },
 
+  "licensesPackageDetailTextZero": "No licenses",
```
Shouldn't we just skip (and not display) any packages that have no licenses at all? That way we don't need to do translation for the zero case. You could update the description to say that it isn't allowed to have a value of zero.

TonicArtos commented:
Sorry, I got caught up this week. I'll make the changes this weekend.

gspencergoog commented:
@TonicArtos have you had a chance to take a look at the changes?

TonicArtos committed:
Refactor build methods into widgets.

- re: https://github.com/flutter/flutter/pull/57588#discussion_r428931325

TonicArtos committed:
Allow omission of localised zero case by returning null by default.

Co-authored-by: Greg Spencer <gspencergoog@users.noreply.github.com>

TonicArtos committed:
Zero case unneeded so removed localised entry.

- re: https://github.com/flutter/flutter/pull/57588#discussion_r430515379

TonicArtos committed:
Fix selection id being lost on ui layout change.

- Caused by refactor into widgets in #f5d3b6bc1c

TonicArtos committed:
Generate localisations.

TonicArtos committed:
Add todo to remove workaround for #59741.

- re: https://github.com/flutter/flutter/pull/57588#discussion_r428935343

TonicArtos commented:
Sorry for the delay. It is a pretty messed up time. Anyway, I don't understand why customer_testing-windows failed all of a sudden. The issue looks like a image pixel comparison check. Is there some way to get more detail on this so I can fix it?

gspencergoog commented on packages/flutter/lib/src/material/about.dart:
```diff
@@ -472,6 +482,314 @@ class LicensePage extends StatefulWidget {
+            style: Theme.of(context).textTheme.bodyText2,
+            textAlign: TextAlign.center,
+          ),
+          Container(height: 18.0),
```
Here, and the next `Container`, please define this constant as a separate static const variable to make it self documenting.

Also, consider using a `SizedBox` instead of a `Container`: `SizedBox` can be a const object, and so it's more performant.

TonicArtos replied:
Thanks.

gspencergoog commented on packages/flutter/lib/src/material/about.dart:
```diff
@@ -623,3 +982,651 @@ Widget _defaultApplicationIcon(BuildContext context) {
+  nested,
+}
+
+const String _navMaster = 'master';
```
These can be `static` strings.

gspencergoog commented on packages/flutter/lib/src/material/about.dart:
```diff
@@ -623,3 +982,651 @@ Widget _defaultApplicationIcon(BuildContext context) {
+    }
+  }
+
+  Widget _nestedUI(BuildContext context) {
```
Consider breaking out this function, and `_lateralUI` into separate `StatelessWidget`s.

gspencergoog commented on packages/flutter/lib/src/material/about.dart:
```diff
@@ -623,3 +982,651 @@ Widget _defaultApplicationIcon(BuildContext context) {
+      actionBuilder: widget.actionBuilder ?? (_, __) => const<Widget>[],
+      automaticallyImplyLeading: widget.automaticallyImplyLeading,
+      centerTitle: widget.centerTitle,
+      detailPageBuilder: (BuildContext context, Object args, ScrollController sc) => widget.detailPageBuilder(context, args ?? _cachedDetailArguments, sc),
```
Use a more descriptive name for `sc`.

gspencergoog commented on packages/flutter/lib/src/material/about.dart:
```diff
@@ -623,3 +982,651 @@ Widget _defaultApplicationIcon(BuildContext context) {
+        return LayoutBuilder(
+            builder: (BuildContext context, BoxConstraints constraints) {
+              final double availableWidth = constraints.maxWidth;
+              if (availableWidth >= (widget.breakpoint ?? 840)) {
```
Make 840 into a constant with a name.

TonicArtos replied:
Thanks.

gspencergoog commented on packages/flutter/lib/src/material/material_localizations.dart:
```diff
@@ -831,6 +834,18 @@ class DefaultMaterialLocalizations implements MaterialLocalizations {
+  @override
+  String licensesPackageDetailText(int licenseCount) {
+    switch (licenseCount) {
+      case 0:
```
If you skip packages with zero licenses, this case should assert.

TonicArtos replied:
Added assertion licenseCount >= 1, removed case 0.

gspencergoog commented on packages/flutter/lib/src/material/about.dart:
```diff
@@ -472,6 +482,314 @@ class LicensePage extends StatefulWidget {
+  String firstPackage;
+
+  void addLicense(LicenseEntry entry) {
+    for (final String package in entry.packages) {
```
Perhaps skip packages with zero licenses?

TonicArtos replied:
I said below in https://github.com/flutter/flutter/pull/57588#discussion_r443198697 that I would do this. However, going over the code again it really is already the case that each package always has one or more licenses. This is because we only know about a package should it be recorded as a package to which a given license applies.

Obviously the code must not be clear enough as is, so I'll go through this part and added comments to try and clarify it.

gspencergoog commented on packages/flutter/lib/src/material/about.dart:
```diff
@@ -472,6 +482,314 @@ class LicensePage extends StatefulWidget {
+            return -1;
+          }
+          if (b == firstPackage) {
+            return -1;
```
Doesn't one of these need to return `1` to keep the first package at the front? Seems like if it returns -1 for both a and b being the first package that it'll be ordered arbitrarily.

TonicArtos replied:
Of course. If b is firstPackage, then a > b, therefore return 1. Thanks.

gspencergoog commented on packages/flutter_localizations/lib/src/l10n/generated_material_localizations.dart:
```diff
@@ -140,6 +140,15 @@ class MaterialLocalizationAf extends GlobalMaterialLocalizations {
+  String get licensesPackageDetailTextOther => '\$licenseCount licenses';
+
+  @override
+  String get licensesPackageDetailTextZero => 'No licenses';
```
All of these "Zero" cases would be eliminated if you skip packages with zero licenses.

TonicArtos replied:
Well, the license page is an inversion of the data hierarchy. LicenseRegistry collects licenses and includes the packages from which those licenses came from as a property on each license. As such, there is a many-to-many relationship between licenses and packages with the constraint that each package has at least one license, and each license has at least one package.

So the zero case is not checked for, since by implementation, it should come up. However, it is undeniably more robust to add checking for, and the exclusion of, packages that have zero licenses. _So I will do this._

However. I would like to highlight, should you include special dispensation for zero license packages, it may be arguable that it is better to highlight packages with zero licenses by using a special unlicensed, or commons, page. But, I also suggest that it is unneeded and zealous work to add this. We should never normally encounter packages with zero licenses. Unless, of course, there is a design change down the line.

TonicArtos replied:
I just realised why this zero case keeps getting regenerated. It is coming from the default text generated for the different locales from when I ran gen_missing_localizations.dart as instructed to do when adding to material_localizations.

gspencergoog reviewed (commented):
Thanks for the update!  It's looking pretty good, I just saw some small things to address.

gspencergoog commented:
As for the `customer_testing` failures, I don't think those were related to your code.  If you rebase the the HEAD of the master branch, I think it will pass.

TonicArtos committed:
Fix sort of packages to keep app package at front.

- re: https://github.com/flutter/flutter/pull/57588#discussion_r442961384

TonicArtos committed:
In default localisation. Assert license count >= 1.

- re: https://github.com/flutter/flutter/pull/57588#discussion_r442955057

TonicArtos commented on packages/flutter/lib/src/material/about.dart:
```diff
@@ -472,6 +482,314 @@ class LicensePage extends StatefulWidget {
+
+  Widget _packageLicensePage(BuildContext _, Object args, ScrollController sc) {
+    assert(args is _DetailArguments);
+    final _DetailArguments a = args as _DetailArguments;
```
Sorry. It looks like I was trying to avoid the stuttering caused by the cast. The style guide should have been followed.

TonicArtos commented on packages/flutter/lib/src/material/about.dart:
```diff
@@ -623,3 +982,651 @@ Widget _defaultApplicationIcon(BuildContext context) {
   return null;
 }
+
+double _getGutterSize(BuildContext context) => MediaQuery.of(context).size.width >= 720 ? 24 : 12;
```
You're right. I remember thinking I would fix this when I wrote it. I clearly didn't.

TonicArtos commented on packages/flutter/lib/src/material/about.dart:
```diff
@@ -556,49 +864,100 @@ class _LicensePageState extends State<LicensePage> {
-                      ),
-                    ),
-                ],
+    final MaterialLocalizations localisations = MaterialLocalizations.of(context);
```
Ah, the z. It kills me. Yeah, okay, I should have stuck to the US spelling no matter how it does the head in.

TonicArtos commented on packages/flutter/lib/src/material/about.dart:
```diff
@@ -472,6 +482,314 @@ class LicensePage extends StatefulWidget {
+    );
+  }
+
+  Widget _packageLicensePage(BuildContext _, Object args, ScrollController sc) {
```
Thanks.

TonicArtos committed:
Clarify generating package and license data.

- Also, move sorting of packages to do a single time, after all
  generation is complete.

TonicArtos committed:
Gb -> US spelling.

- re: https://github.com/flutter/flutter/pull/57588#discussion_r442945087

TonicArtos committed:
Use sized box instead of container for spacing.

- re: https://github.com/flutter/flutter/pull/57588#discussion_r442951229

TonicArtos committed:
Do not use abbreviations.

- re: https://github.com/flutter/flutter/pull/57588#discussion_r442950390

TonicArtos committed:
Do not use magic numbers.

- re: https://github.com/flutter/flutter/pull/57588#discussion_r442947725
- re: https://github.com/flutter/flutter/pull/57588#discussion_r442954666

TonicArtos commented:
So I did the rebase. See if this will work. I can redo it with a squash if you like.

TonicArtos committed:
Update l10n test to match rejected zero case.

I am not entirely sure this is the right way to go about this.

TonicArtos committed:
Use throwsAssertionError instead of matcher.

gspencergoog commented on packages/flutter/test/material/about_test.dart:
```diff
@@ -225,7 +256,12 @@ void main() {
+    // the safe area is sufficiently respected.
+    expect(
+      tester.getTopLeft(find.text('Licenses')),
+      const Offset(16.0 + safeareaPadding, 18 + safeareaPadding),
```
Be consistent about doubles: either use `16`, or `18.0`.

TonicArtos replied:
Okay, I'll go through the change list and conform to `18.0` everywhere. On closer inspection it seems to be the convention.

gspencergoog reviewed (approved):
This looks pretty good!  And it looks like the customer_testing is passing now too.

![32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d](https://user-images.githubusercontent.com/8867023/37737237-66e9c9dc-2d10-11e8-8797-13a405eace49.png)

Unless you have any more updates, I can merge this as soon as the tree is green.  I'll wait until you say it's ready, though.

TonicArtos committed:
Append .0 to doubles to distinguish from ints.

- re: https://github.com/flutter/flutter/pull/57588#discussion_r443736276

TonicArtos committed:
Fix incorrect documentation of zero case decl.

TonicArtos commented:
> This looks pretty good! And it looks like the customer_testing is passing now too.
> 

Okay I addressed your last comment. I also went through the CL one last time and noticed the documentation still mentioned the localised zero case was required, so I fixed that too.

> Unless you have any more updates, I can merge this as soon as the tree is green. I'll wait until you say it's ready, though.

I see no reason to hold back, once these last two commits complete the CI process and the tree goes green again.

Oh, if you want to talk about the MDFlow thing and seeing if that can go somewhere, I'd be happy to. I think with work it could become something quite good.

Otherwise, or in addition, I'll move on and see about getting all this done again for the Cupertino side of things too. Over at #57638.

gspencergoog commented:
Thanks Tonic! Great work, we really appreciate it!

I'll commit this as soon as the tree is green and it passes the CI tests.

TonicArtos commented:
@gspencergoog Do you want me to squash the PR or are you going to do that?

gspencergoog commented:
No, no need to squash: it'll do that automatically when I merge. The tree's been extra red: we had an issue in our device lab.

goderbauer commented on packages/flutter/lib/src/material/about.dart:
```diff
@@ -625,3 +1009,658 @@ Widget _defaultApplicationIcon(BuildContext context) {
+    final double minHeight = (screenHeight - kToolbarHeight) / screenHeight;
+
+    return GestureDetector(
+      onTap: () {
+        print('draggable');
+      },
```
This print (and probably the entire gesture detector?) should probably be removed?

gspencergoog replied:
Ooh, yes, good point.  @TonicArtos can you take care of that?

TonicArtos replied:
Yes. It is late here. I'll get caught up with up and get the fix sorted tomorrow.

gspencergoog replied:
OK, no huge rush: most of us will be on holiday until Monday (it's Independence Day in the U.S.).

TonicArtos replied:
I went through my journal. It turns out the GestureDetector is left from when I was looking at #59741. I forgot to remove it once I figured what was happening.

TonicArtos replied:
PR #60730 Remove superfluous GestureDetector.

TonicArtos commented:
Yes. Do you want a new pull request for it?

On Thu, 2 Jul 2020, 04:50 Greg Spencer, <notifications@github.com> wrote:

> *@gspencergoog* commented on this pull request.
> ------------------------------
>
> In packages/flutter/lib/src/material/about.dart
> <https://github.com/flutter/flutter/pull/57588#discussion_r448609107>:
>
> > +      onTap: () {
> +        print('draggable');
> +      },
>
> Ooh, yes, good point. @TonicArtos <https://github.com/TonicArtos> can you
> take care of that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/flutter/pull/57588#discussion_r448609107>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACIJTUTLTMO6LRQH7NQNJ3RZOOR3ANCNFSM4NFHFFJQ>
> .
>

gspencergoog commented:
Sure, that would be great.

mono0926 commented on packages/flutter/lib/src/material/about.dart:
```diff
@@ -558,49 +891,100 @@ class _LicensePageState extends State<LicensePage> {
+    if (widget.scrollController == null) {
+      page = Scaffold(
+        appBar: AppBar(
+          title: _PackageLicensePageTitle(title, subtitle, theme.primaryTextTheme),
```
@TonicArtos 

Why did you specify `theme.primaryTextTheme`?

It may break color system when `AppBarTheme` is used for example.

ingmferrer replied:
Indeed, it broke my background color.

HansMuller replied:
I would have preferred for this to (private) widget to have been factored a little differently - to not have used primaryTextTheme.

If you can file an issue with a simple test case that demonstrates a regression, we'll try and fix it.

mono0926 replied:
@HansMuller 

> If you can file an issue with a simple test case that demonstrates a regression, we'll try and fix it.

Okay, I've created an issue: #63239.

mono0926 replied:
@HansMuller 

I tried to fix at #63249, but I’m welcome to be closed and you enhance the implementation.

TonicArtos replied:
> I would have preferred for this to (private) widget to have been factored a little differently - to not have used primaryTextTheme.
> 
> If you can file an issue with a simple test case that demonstrates a regression, we'll try and fix it.
> 

Sorry about the issue. What do you think the original solution should have been?

mono0926 replied:
@TonicArtos 

> What do you think the original solution should have been?

I think #63239 is the simplest fix for exiting implementation.
